desc/objectives/_geometry.py

```diff
@@ -459,6 +459,290 @@ def compute(self, params, constants=None):
+
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, params, constants=None):
+        """Compute data of coil for given data key.
+
+        Parameters
+        ----------
+        params : dict
+            Dictionary of the coil's degrees of freedom.
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc. Defaults to
+            self._constants.
+
+        Returns
+        -------
+        f : float or array of floats
+            Coil length.
+        """
+        if constants is None:
+            constants = self._constants
+
+        if constants["is_mixed_coils"]:
+            coils = self.things[0].coils
+            data = [
+                coils[i].compute(
+                    self._data_keys,
+                    params=params[i],
+                    transforms=constants["transforms"][i],
+                    grid=self._grid[i],
+                )
+                for i, coil in enumerate(coils)
+            ]
+        else:
+            coils = self.things[0]
+            data = coils.compute(
+                self._data_keys,
+                params=params,
+                transforms=constants["transforms"],
+                grid=self._grid,
+            )
+
+        return data
+
+
+class CoilLength(_CoilObjective):
+    """Coil length.
+
+    Parameters
+    ----------
+    coil : CoilSet or Coil
+        Coil(s) that are to be optimized
+    target : {float, ndarray}, optional
+        Target value(s) of the objective. Only used if bounds is None.
+        Must be broadcastable to Objective.dim_f.
+    bounds : tuple of {float, ndarray}, optional
+        Lower and upper bounds on the objective. Overrides target.
+        Both bounds must be broadcastable to to Objective.dim_f
+    """
+
+    def __init__(self, coil, target=None, bounds=None, **kwargs):
+        if target is None and bounds is None:
+            target = 2 * np.pi
+
+        super().__init__(coil, ["length"], target=target, bounds=bounds, **kwargs)
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, params, constants=None):
+        """Compute coil length.
+
+        Parameters
+        ----------
+        params : dict
+            Dictionary of the coil's degrees of freedom.
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc. Defaults to
+            self._constants.
+
+        Returns
+        -------
+        f : float or array of floats
+            Coil length.
+        """
+        data = super().compute(params, constants=constants)
+        if isinstance(data, list):
```

Comment:
In all cases data here will either be a dict or nested list of dict from possibly nested coilsets, so you can use the `tree_flatten` thing again:

```python
data = tree_flatten(data, is_leaf=lambda x: isinstance(x, dict))[0]
out = jnp.concatenate([dat[key] for dat in data])
```

Reply:
You could also allow the target/bounds etc to be nested pytrees that match the structure of the coilset and use this same thing to flatten them down to 1d arrays